[DSM] PEPPER-873

pegahtah opened the pull request:
## Context

Today during a debug session we realized we are not keeping the message and response we sent and received from Mercury, since these kind of debugs are happening often I added 2 more fields to ddp_mercury_sequencing to keep them

## Checklist

- [x] I have labeled the type of changes involved using the `C-*` labels.
- [ ] I have assessed potential risks and labeled using the `R-*` labels.
- [ ] I have considered error handling and alerts, and added `L-*` labels as needed.
- [ ] I have considered security and privacy, and added `I-*` labels as needed
- [ ] I have analyzed my changes for stability, fault tolerance, graceful degradation, performance bottlenecks and written a brief summary in this PR.
- [ ] If applicable, I have discussed the analytics needs at both a platform and study level with Product and instrumented code accordingly.
- [ ] If applicable, my UI/UX changes have passed muster with Product/Design via an over-the-shoulder review, screenshots, etc.

_If unsure or need help with any of the above items, add the `help wanted` label. For items that starts with `If applicable`, if it is not applicable, check it off and add `n/a` in front._

## FUD Score

_Overall, how are you feeling about these changes?_

- [x] :relaxed: All good, business as usual!
- [ ] :sweat_smile: There might be some issues here
- [ ] :scream: I'm sweaty and nervous

## How do we demo these changes?

_How does one observe these changes in a deployed system? Note that **user visible** encompasses many personas--not just patients and study staff, but also ops duty, your fellow devs, compliance, etc._

- [ ] They are user-visible in dev as a regular user journey and require no additional instructions.
- [ ] Getting dev into a state where this is user-visible requires some tech fiddling. I have documented these steps in the related ticket.
- [ ] Requires other features before it's human visible. I have documented the blocking issues in jira.
- [ ] I have no idea how to demo this. Please help me!

## Testing

- [ ] I have written automated positive tests
- [ ] I have written automated negative tests
- [x] I have written zero automated tests but have poked around locally to verify proper functionality
- [ ] The jira ticket has acceptance criteria and QA has the needed information to test changes

## Release

- [x] These changes require no special release procedures--just code!
- [ ] Releasing these changes requires special handling and I have documented the special procedures in the release plan document

pegahtah committed:
PEPPER-873
adding 2 more fields so DSM tracks what is sent to Mercury and what is received as result
Adding more logs

pegahtah commented on pepper-apis/dsm-core/src/main/resources/liquibase/PEPPER-873_mercury_order_fields.xml:
```diff
@@ -0,0 +1,17 @@
+        </addColumn>
+    </changeSet>
+    <changeSet id="PEPPER-873_mercury_status_detail_changes" author="pegah">
+        <modifyDataType columnName="status_detail" newDataType="VARCHAR(2000)" tableName="ddp_mercury_sequencing"/>
```
There are errors in dev that the field is small so I changed the size

otchet-broad replied:
Why not convert the field to text?

pegahtah replied:
@otchet-broad if our theory is correct I think we will see a size increase in the messages again

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/mercury/MercuryOrderDao.java:
```diff
@@ -171,10 +173,19 @@ public HashMap<String, MercuryOrderDto> getPossibleBarcodesForParticipant(String
         });
+
+        if (results.resultException != null) {
+            throw new RuntimeException("Error getting the mercury orders", results.resultException);
```
What will catch and process this exception?
Is updateOrderStatus always called from a pubsub message processor?

otchet-broad replied:
@denniscunningham - In this pattern, we seem to catch and swallow `SQLException` within the transaction scope at line 168/170.  If I'm understanding the goal of this change, it's to gain more insight into the errors being encountered here that aren't currently bubbling up.

@pegahtah  - Please confirm.

denniscunningham replied:
Thanks Elliot. But where exactly will this exception be caught and reported?

otchet-broad replied:
If I understand our codebase correctly, this will bubble all the way up to and get logged via the RequestHandler path through Spark.

denniscunningham replied:
Thx. I read the call stack incorrectly earlier. I see the route now.

Is there translation of RuntimeException to a specific HTTP status code in Spark? 
In this case, is this a request error or an internal error?

otchet-broad replied:
Any that aren't handled directly are turned into a 500.  In this case, it will present as a 500 to the user.

denniscunningham replied:
Is this an internal error or a problem with the request?

otchet-broad replied:
The best answer for this case is: yes - it's either a request error or an internal one.  It's likely a case of a bad request, but there's at least one edge case I can see that bubbles up where it wouldn't be: the database goes down during this part of the request/response lifecycle.
The only thing this addition is doing is surfacing the error instead of returning a possibly incomplete map.
What would you like to see happen here?  We could continue with the pattern as-is, change it for this case, or change it for all of the cases.

denniscunningham replied:
Thx for the explanation Elliot.
I'd like us to return an appropriate response code from all our routes, but I'm starting to realize that's a big job. So for now if we know the code we want to return (for code under change or development) we should try to align exceptions with response codes. (I see this, but I am not sure if there is a better approach: http://sparkjava.com/documentation.html#error-handling. I have not used Spark before -- at least not this Spark.)
If there is not a clean opportunity to quickly fix those issues here, of course go ahead with the merge. I appreciate what Pegah is trying to accomplish with the PR.

otchet-broad commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/mercury/MercuryOrderDao.java:
```diff
@@ -171,10 +173,19 @@ public HashMap<String, MercuryOrderDto> getPossibleBarcodesForParticipant(String
 
+    @Override
+    public int create(MercuryOrderDto mercuryOrderDto) {
+        return 0;
```
Should this really throw something like ("Not implemented") instead of returning 0?

otchet-broad commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/pubsub/MercuryOrderPublisher.java:
```diff
@@ -99,6 +91,15 @@ public void onSuccess(String messageId) {
     }
 
+    private String createMercuryUniqueOrderId() {
+        String orderNumber = NanoIdUtil.getNanoId("1234567890QWERTYUIOPASDFGHJKLZXCVBNM", 20);
```
What's the volume of orders here?  
How often will we have ID collision?
Are these for human consumption?  If not, how about a UUID?

pegahtah replied:
I think the volume is ~ 10 orders every month or so

pegahtah committed:
PEPPER-873
some fixes for the comparison method in ClinicalOrderDao.java and the test for it in ComparisonTest.java
Also adding an exception to the overwritten `create` method in MercuryOrderDao.java

otchet-broad reviewed (approved):
Thanks Pegah!

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into PEPPER-873

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into PEPPER-873

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into PEPPER-873

# Conflicts:
#	pepper-apis/dsm-core/src/main/resources/master-changelog.xml

pegahtah committed:
Merge branch 'develop' into PEPPER-873
